packages/dart_firebase_admin/lib/src/utils/crypto_signer.dart

```diff
@@ -107,11 +107,33 @@ class _ServiceAccountSigner implements CryptoSigner {
 
   @override
   Future<Uint8List> sign(Uint8List buffer) async {
-    final key = utf8.encode(credential.privateKey);
-    final hmac = Hmac(sha256, key);
-    final digest = hmac.convert(buffer);
+    final rsaPrivateKey = _parsePrivateKeyFromPem();
+    final signer = Signer('SHA-256/RSA')
+      ..init(true, PrivateKeyParameter<RSAPrivateKey>(rsaPrivateKey));
+    final signature = signer.generateSignature(buffer) as RSASignature;
+    return signature.bytes;
+  }
+
+  RSAPrivateKey _parsePrivateKeyFromPem() {
+    final privateKeyString = credential.privateKey
+        .replaceAll('-----BEGIN PRIVATE KEY-----', '')
+        .replaceAll('-----END PRIVATE KEY-----', '')
+        .replaceAll('\n', '');
+    final privateKeyDER = base64Decode(privateKeyString);
+
+    final asn1Parser = ASN1Parser(Uint8List.fromList(privateKeyDER));
+    final topLevelSequence = asn1Parser.nextObject() as ASN1Sequence;
```

Comment:
Too many casts and `!` in there. Could you remove those? Such as by using `is` and `!= null`

We'd also need tests for when those values are invalid

Reply:
@rrousselGit 

Thank you! I'd like you to give some advice about this point!

I'm thinking to improve the code like the following to reduce `as` casting and `!`:

<details>

<summary>crypto_signer.dart</summary>

```dart
RSAPrivateKey _parsePrivateKeyFromPem() {
  final privateKeyString = credential.privateKey
      .replaceAll('-----BEGIN PRIVATE KEY-----', '')
      .replaceAll('-----END PRIVATE KEY-----', '')
      .replaceAll('\n', '');
  final privateKeyDER = base64Decode(privateKeyString);

  final asn1Parser = ASN1Parser(Uint8List.fromList(privateKeyDER));
  final topLevelSequence = asn1Parser.nextObject();
  if (topLevelSequence is! ASN1Sequence) {
    throw const FormatException('Invalid ASN1 format for private key');
  }
  final privateKeyOctet = topLevelSequence.elements?[2];
  if (privateKeyOctet is! ASN1OctetString) {
    throw const FormatException('Invalid ASN1 format for private key');
  }

  final privateKeyParser = ASN1Parser(privateKeyOctet.valueBytes);
  final privateKeySequence = privateKeyParser.nextObject();
  if (privateKeySequence is! ASN1Sequence) {
    throw const FormatException('Invalid ASN1 format for private key');
  }

  final modulus = privateKeySequence.elements?[1];
  final exponent = privateKeySequence.elements?[3];
  final p = privateKeySequence.elements?[4];
  final q = privateKeySequence.elements?[5];
  if (modulus is! ASN1Integer ||
      exponent is! ASN1Integer ||
      p is! ASN1Integer ||
      q is! ASN1Integer) {
    throw const FormatException('Invalid ASN1 format for private key');
  }

  return RSAPrivateKey(
    modulus.integer!,
    exponent.integer!,
    p.integer,
    q.integer,
  );
}
```
</details>

and tried to write the unit test for success case like the following:

<details>

<summary>crypto_signer.dart</summary>

```dart
test('parsePrivateKeyFromPem should parse valid private key', () {
  final credential = Credential.fromServiceAccountParams(
    clientId: 'id',
    privateKey: _fakeRSAKey,
    email: 'email',
  );
  final app = FirebaseAdminApp.initializeApp('project-id', credential);
  final signer = cryptoSignerFromApp(app);
  expect(signer, isA<CryptoSigner>());
  expect(signer.algorithm, 'RS256');

  final buffer = Uint8List.fromList(List.generate(32, (i) => i));

  expect(() async => signer.sign(buffer), returnsNormally);
});
```

</details>

However, when I try to write test to confirm the code throwing the  `FormatException`, I need to give invalid private key string to `Credential.fromServiceAccountParams`, and it throws Invalid DER encoding error before reaching `_ServiceAccountSigner.sign` (`_ServiceAccountSigner._parsePrivateKeyFromPem`) method.

I appreciate if you have any ideas about it!

Reply:
One way to do it is to make `_ServiceAccountSigner` "public", but have `export ... hide ServiceAccountSigner`.

Make sure to make `ServiceAccountSigner` with `@internal` as you do so.